src/arm/windows/init_by_logical_sys_info.c

```diff
@@ -0,0 +1,842 @@
+		(uint16_t) group_id;
+	processors[processor_global_index].windows_processor_id =
+		(uint16_t) processor_id_in_group;
+
+	/* As we're counting the number of packages now, we haven't allocated memory for */
+	/* cpuinfo_packages yet, so we only set the package pointer's offset now. */
+	processors[processor_global_index].package =
+		(const struct cpuinfo_package*) NULL + package_id;
+}
+
+void store_core_info_per_processor(
+	struct cpuinfo_processor* processors,
+	const uint32_t processor_global_index,
+	const uint32_t core_id,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX core_info,
+	struct cpuinfo_core* cores,
+	const struct woa_chip_info *chip_info)
+{
+	if (cores) {
+		processors[processor_global_index].core = cores + core_id;
+		cores[core_id].core_id = core_id;
+		get_core_uarch_for_efficiency(
+			chip_info->chip_name, core_info->Processor.EfficiencyClass,
+			&(cores[core_id].uarch), &(cores[core_id].frequency));
+
+		/* We don't have cluster information, so we handle it as
+		 * fixed 1 to (cluster / cores).
+		 */
+		processors[processor_global_index].cluster =
+			(const struct cpuinfo_cluster*) NULL + core_id;
+	}
+}
+
+static void store_cache_info_per_processor(
+	struct cpuinfo_processor* processors,
+	const uint32_t processor_global_index,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX info,
+	struct cpuinfo_cache* current_cache)
+{
+	if (current_cache->processor_start > processor_global_index) {
+		current_cache->processor_start = processor_global_index;
+	}
+	current_cache->processor_count++;
+
+	switch(info->Cache.Level) {
+		case 1:
+			switch (info->Cache.Type) {
+				case CacheInstruction:
+					processors[processor_global_index].cache.l1i = current_cache;
+				break;
+				case CacheData:
+					processors[processor_global_index].cache.l1d = current_cache;
+				break;
+			}
+		break;
+		case 2:
+			processors[processor_global_index].cache.l2 = current_cache;
+		break;
+		case 3:
+			processors[processor_global_index].cache.l3 = current_cache;
+		break;
+	}
+}
+
+static void connect_packages_cores_clusters_by_processors(
+	struct cpuinfo_processor* processors,
+	const uint32_t nr_of_processors,
+	struct cpuinfo_package* packages,
+	const uint32_t nr_of_packages,
+	struct cpuinfo_cluster* clusters,
+	struct cpuinfo_core* cores,
+	const uint32_t nr_of_cores,
+	const struct woa_chip_info* chip_info,
+	enum cpuinfo_vendor vendor)
+{
+	/* Adjust core and package pointers for all logical processors. */
+	for (uint32_t i = nr_of_processors; i != 0; i--) {
+		const uint32_t processor_id = i - 1;
+		struct cpuinfo_processor* processor = processors + processor_id;
+
+		struct cpuinfo_core* core = (struct cpuinfo_core*)processor->core;
+
+		/* We stored the offset of pointers when we haven't allocated memory
+		 * for packages and clusters, so now add offsets to base addresses.
+		 */
+		struct cpuinfo_package* package =
+			(struct cpuinfo_package*) ((uintptr_t) packages + (uintptr_t) processor->package);
+		processor->package = package;

```

Comment:
this needs a sanity check, i.e. processor->package should not be less than it's original value (overflow) and not be greater than the total size of packages.